packages/@aws-cdk/aws-lambda/lib/function-base.ts

```diff
@@ -246,12 +246,27 @@ export abstract class FunctionBase extends cdk.Construct implements IFunction  {
   /**
    * Grant the given identity permissions to invoke this Lambda
    */
-  public grantInvoke(identity?: iam.IPrincipal) {
-    if (identity) {
-      identity.addToPolicy(new iam.PolicyStatement()
-        .addAction('lambda:InvokeFunction')
-        .addResource(this.functionArn));
-    }
+  public grantInvoke(principal?: iam.IPrincipal): iam.Grant {
+    return iam.Permissions.grantWithResource({
+      principal,
+      actions: ['lambda:InvokeFunction'],
+      resourceArns: [this.functionArn],
+
+      // Fake resource-like object on which to call addToResourcePolicy(), which actually
+      // calls addPermission()
+      resource: {
+        addToResourcePolicy: (_statement) => {
```

Comment:
I know.

I considered doing that (in fact, we have a ticket about it), but it's tricky in its generality. There's only a subset of the full IAM spec support by the `Permission` model, mostly around `Conditions`. So exposing `addToResourcePolicy(statement)` publicly will not work for all `statement`s, and in the face of tokens I don't even know whether it's necessarily possible do determine whether the statement will work or not.

The nice solution of doing it this way is that when `addToResourcePolicy()` is called, we know exactly what it's trying to do, and it won't involve any conditions (or more correctly, we know the conditions it will be trying to add), so it's possible to do the translation.